File: shared/src/commonMain/kotlin/com/darskhandev/urail/viewmodel/HomeViewModel.kt
package com.darskhandev.urail.viewmodel

import com.darskhandev.urail.data.source.remote.models.pathfinder.Graph
import com.darskhandev.urail.data.source.remote.models.pathfinder.findNearestStation
import com.darskhandev.urail.data.source.remote.models.schedule.DetailScheduleRaw
import com.darskhandev.urail.data.source.remote.models.schedule.RouteScheduleRaw
import com.darskhandev.urail.domain.entities.Route
import com.darskhandev.urail.domain.entities.Schedule
import com.darskhandev.urail.domain.entities.Station
import com.darskhandev.urail.domain.entities.StationList
import com.darskhandev.urail.domain.usecases.UseCases
import com.darskhandev.urail.utils.AppState
import com.darskhandev.urail.utils.GetRouteSchedule
import com.darskhandev.urail.utils.RouteSearch
import com.darskhandev.urail.utils.ScheduleSearch
import dev.icerock.moko.mvvm.viewmodel.ViewModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.transform
import kotlinx.coroutines.flow.transformLatest
import org.koin.core.component.KoinComponent

class HomeViewModel(
    private val useCases: UseCases,
) : ViewModel(), KoinComponent {
    private val _routeState = MutableStateFlow<AppState<Route>>(AppState.NoState())
    val routeState = _routeState.asStateFlow()
    private val _routeScheduleState =
        MutableStateFlow<AppState<List<RouteScheduleRaw>>>(AppState.NoState())
    val routeScheduleState = _routeScheduleState.asStateFlow()

    private val initialScheduleSearch = ScheduleSearch("", "", "")
    private val scheduleSearch = MutableStateFlow(initialScheduleSearch)

    private val initialGetRoute = RouteSearch("", "", "")
    private val getRouteState = MutableStateFlow(initialGetRoute)
    fun getScheduleSearch(stationId: String, timeFrom: String, timeTo: String) {
        scheduleSearch.value = ScheduleSearch(stationId, timeFrom, timeTo)
    }

    private val kaIdforDetailSchedule = MutableStateFlow("")
    fun setKaIdforDetailSchedule(kaId: String) {
        kaIdforDetailSchedule.value = kaId
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    val getDetailSchedule = kaIdforDetailSchedule.distinctUntilChanged { old, new -> old == new }
        .transformLatest { it ->
            if (it == "") {
                emit(AppState.Error("Please select station and time"))
            } else {
                useCases.getDetailSchedule(it).collect {
                    this.emit(it)
                }
            }
        }

    @OptIn(ExperimentalCoroutinesApi::class)
    val getSchedule = scheduleSearch.distinctUntilChanged { old, new -> old == new }
        .transformLatest { it ->
            if (it == initialScheduleSearch) {
                emit(AppState.Error("Pilih Stasiun dan waktu terlebih dahulu"))
            } else {
                useCases.getKrlSchedule(it.stationId, it.timeFrom, it.timeTo).collect {
                    this.emit(it)
                }
            }
        }

    fun setRouteSearch(originStationId: String, destinationStationId: String, departureTime: String) {
        getRouteState.value = RouteSearch(originStationId, destinationStationId, departureTime)
    }

    fun findRoute(originLat: Double, originLng: Double, destinationLat: Double, destinationLng: Double, departureTime: String) {
        val originStationId = getGraph.value.data?.findNearestStation(originLat, originLng)?.stationId
        val destinationStationId = getGraph.value.data?.findNearestStation(destinationLat, destinationLng)?.stationId
        if (originStationId == null || destinationStationId == null) {
            _routeState.value = AppState.Error("Anda Sepertinya berada di luar area KRL")
            return
        }
        getRouteState.value = RouteSearch(originStationId, destinationStationId, departureTime)
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    val getRoute = getRouteState.distinctUntilChanged { old, new -> old == new }
        .transformLatest { it ->
            if (it == initialGetRoute) {
                emit(AppState.Error("Ayo Berangkat"))
            } else {
                useCases.getKrlRoute(it.originStationId, it.destinationStationId, it.departureTime)
                    .collect {
                        this.emit(it)
                    }
            }
        }

    val getStationList: StateFlow<AppState<StationList>> =
        useCases.getKrlStations().stateIn(
            viewModelScope,
            SharingStarted.Lazily,
            AppState.NoState()
        )

    val getGraph: StateFlow<AppState<Graph>> =
        useCases.getGraph().stateIn(
            viewModelScope,
            SharingStarted.Lazily,
            AppState.NoState()
        )

    private val initialGetRouteSchedule = GetRouteSchedule(emptyList(), "")
    private val getRouteScheduleState = MutableStateFlow(initialGetRouteSchedule)
    fun setRouteSchedule(routesId: List<String>, timeFrom: String) {
        getRouteScheduleState.value = GetRouteSchedule(routesId, timeFrom)
    }
    @OptIn(ExperimentalCoroutinesApi::class)
    val getRouteSchedule = getRouteScheduleState.distinctUntilChanged{ old, new -> old == new }
        .transformLatest { it ->
            if (it == initialGetRouteSchedule) {
                emit(AppState.Error("Ayo Berangkat"))
            } else {
                useCases.getRouteSchedule(it.routes, it.timeFrom).collect {
                    this.emit(it)
                }
            }
        }
}

File: iosApp/iosApp/presentation/viewmodels/HomeViewmodel.swift
//
//  HomeViewmodel.swift
//  iosApp
//
//  Created by Akashaka on 19/06/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

public class HomeObservableObject: ObservableObject {
    static let shared = GetComponents().getHomeViewModel().asObservableObject()
    var viewModel: HomeViewModel
    @Published var graphState: AppState<Graph> = AppStateNoState(message: "initial State")

    @Published var stationsState: AppState<StationList> = AppStateNoState(message: "initial State")
    var stationsStateStrings: [String] {
        return (stationsState.data?.stations ?? []).map { station in
            station.stationName
        }
    }

    var stationsIdState: [String] {
        return (stationsState.data?.stations ?? []).map { station in
            station.stationId
        }
    }

    @Published var detailScheduleState: AppState<DetailScheduleRaw> = AppStateNoState(message: "initial State")
    @Published var scheduleState: AppState<ScheduleList> = AppStateNoState(message: "initial State")
    @Published var routeState: AppState<Route> = AppStateNoState(message: "initial State")
    @Published var routeScheduleState: AppState<ListRouteScheduleRaw> = AppStateNoState(message: "initial State")

    init(wrapper: HomeViewModel) {
        viewModel = wrapper
        print("init")
        // get stations
        viewModel.getStationList.collect(collector: Collector<AppState<StationList>> { state in
            if let stationList = state.data {
                TimelineState.shared.stations = stationList.stations
            }
            self.stationsState = state
        }) { error in
            self.stationsState = AppStateError(error: error?.localizedDescription ?? "Unknown Error", data: nil)
        }
        // Map Feature
        viewModel.getGraph.collect(collector: Collector<AppState<Graph>> { state in
            self.graphState = state
        }) { error in
            self.graphState = AppStateError(error: error?.localizedDescription ?? "Unknown Error", data: nil)
        }

        // schedule feature
        viewModel.getSchedule.collect(collector: Collector<AppState<ScheduleList>> { state in
            self.scheduleState = state
            print(state)
        }) { error in
            print(error)
            self.scheduleState = AppStateError(error: error?.localizedDescription ?? "Unknown Error", data: nil)
        }
        viewModel.getDetailSchedule.collect(collector: Collector<AppState<DetailScheduleRaw>> { state in
            self.detailScheduleState = state
        }) { error in
            self.detailScheduleState = AppStateError(error: error?.localizedDescription ?? "Unknown Error", data: nil)
        }
        // route feature
        viewModel.getRoute.collect(collector: Collector<AppState<Route>> { state in
            if state is AppStateSuccess {
                print(state)
            }
            self.routeState = state
        }) { error in
            self.routeState = AppStateError(error: error?.localizedDescription ?? "Unknown Error", data: nil)
        }
        // route schedule feature
        viewModel.getRouteSchedule.collect(collector: Collector<AppState<ListRouteScheduleRaw>> { state in
            if state is AppStateSuccess {
                print(state)
            }
            self.routeScheduleState = state
        }) { error in
            self.routeScheduleState = AppStateError(error: error?.localizedDescription ?? "Unknown Error", data: nil)
        }
    }

    func findSchedule(stationId: String, timeFrom: String, timeTo: String) {
        viewModel.getScheduleSearch(stationId: stationId, timeFrom: timeFrom, timeTo: timeTo)
    }

    func setKaIdDetailSchedule(kaId: String) {
        viewModel.setKaIdforDetailSchedule(kaId: kaId)
    }

    func findRoute(originStationId: String, destinationStationId: String, departureTime: String) {
        viewModel.setRouteSearch(originStationId: originStationId, destinationStationId: destinationStationId, departureTime: departureTime)
    }

    func findRouteByLocation(start: LocationPoint, destination: LocationPoint, departTime: String) {
        viewModel.findRoute(originLat: start.latitude, originLng: start.longitude, destinationLat: destination.latitude, destinationLng: destination.longitude, departureTime: departTime)
    }

    func loadRouteSchedule(routes: [String], timeFrom: String) {
        print(routes)
        print(timeFrom)
        viewModel.setRouteSchedule(routesId: routes, timeFrom: timeFrom)
    }
}

public extension HomeViewModel {
    func asObservableObject() -> HomeObservableObject {
        return HomeObservableObject(wrapper: self)
    }
}
